### PR TITLE
BoardNotes fixed to work on Kanboard v1.2.33 and PHP 8.2 + multiple fixes and enhancements + heavy refactoring

### PR DESCRIPTION
**Pull request by imfx77**
I initially adopted this project as a replacement for another plugin that was not working nor maintained (Wiki) as I hoped that it could cover some functionality I was missing in Kanboard at least partially. To my disappointment BoardNotes didn't work either, had multiple crashes and seemed to be abandoned amidst some general rework. But kinda looked more simple and easy to pick up and fix.

So I intended to just make it work in order to consider if it will do the job for me. But then started fixing here and there, reworking stale or ugly code, small enhancement here, another bigger there, sanitizing or unifying code ... and after a short time it was clear that it wasn't going to be a small rework. As you can see, it ended up in quite a bit of revamping all code, visuals, and concept.

And I would totally understanf if you don't bother to review and merge this pull request as it is huge. And furthermore, it changes and introduces so many new things that it may easily derail from your original vision of the plugin. Please, let me know if so - then I can consider keeping the fork of the project for my personal use (or for someone else that may find my upgrade useful) and eventually maintaining it further. I avoided updating documentation, major plugin version, or version history before I know which way it shall go.

Fully descibing the changes would be actually quite some additional effort. I recommend that you just checkout and try the plugin for yourself - and see if it looks and works alright for you, so to consider if you like or not my take on the matter. Code changes are propably too hard to follow, so again, I would recommend viewing the general picture as code style, naming convention, architecture, arrangement, etc. Of course, you can dive into each of my 40+ commits, and you can trace more closely what exactly has changed and how. And why ;)
As I am veeery thorough in the comments.

Thank you for your time!
And have nice holidays :)

### CODE REVIEW

**Comment by aljawaid** [comment]
Hi this is great, I will look into it very soon.

I took over the project for the same reason as you I think but stumbled across a lot of old code which slowed me down.

Your work has motivated to get this released properly which I will do soon.

I will be in touch.

**Comment by imfx77** [comment]
@aljawaid 
There are 4 new commits on my branch that were fairly important and necessary.
( 2x hotfixes of major behavior bugs + small enhancement + styles adjustments for readability )
As usually accompanied by code cleanup, refactoring and unification.
Sorry about that ;/
But now that those are wrapped and finished - I finally have a fairly functional and polished plugin which I can work with.
And I will restrain myself from pushing further commits until you manage to assess all the changes.

**Comment by aljawaid** [comment]
@imfx77 I have merged this, will do the testing on my side

**Comment by aljawaid** [comment]
The problem that delayed #2 to making this plugin reach an initial release was that the original developer coded using standard php, whereas Kanboard requires PSR-1 style coding.  I guess if it's working, it is working but it slows me down a lot to understand the code.

**Comment by imfx77** [comment]
Well, to be honest I am no web developer at all - first time seriously diving into PHP + JS + CSS.
But I am a 20+ years C++ dev, worked on some quite large scale and complex projects - and I can definitely recognize when one is incoherent, no matter the language or the coding style or the paradigm usage or the methodology involved.
I didn't have any idea what style coding should be used in a KB plugin. As I mentioned, I just aimed for a quick fix, wasn't going to get involved so deep. But as I began reworking anyways, I mostly looked at how other KB plugins are implemented and also aimed at better code modularity isolation and readability from a pure common sense. Yet, I restrained myself to minimize the changes, so that it was easier to review ... which obviously didn't work as intended ;/

I should have guessed it is not your original project. Now, that KB development is almost stalled, and most new functionality and enhancements come through side-dev plugins - I can recognize you as one of the most active contributors, and I'm using myself several of your plugins - and BoardNotes was NOT made in the same way. So I should know you adopted it somewhere on the road. And if I knew how you felt about about reworking it prior to reworking it ... I would only took the idea and started it from scratch rather than trying to relate to the original code. In my **dev** branch I continue the rework and enhancements, and anyways in the end the structure and code will be just vaguely reminding of the "original".

And I don't wanna be giving you constant headaches reviewing and merging massive rework changes.
So maybe consider giving the whole plugin a new beginning? Or even make it a fresh new plugin?
Pls, let me know what you think about that.

**Comment by aljawaid** [comment]
Hello....

> Yet, I restrained myself to minimize the changes, so that it was easier to review ... which obviously didn't work as intended ;/
- the coding issue I was having with this plugin starts from the initial way it was coded. Not your changes.

> I should have guessed it is not your original project.
- I need a post-it style system for 'temporary' notes related to a project/task which is why I found this and offered to take it over.  I should've gone through the code first, but I thought it was a working plugin - it was but for older KB versions without databases.

> KB development is almost stalled, and most new functionality and enhancements come through side-dev plugins
- yes it has been relativly stalled for over a year now, but the plugins and community do keep it alive.  The issue is, there is no other replacement with the same set of features - everything else I found was commercial or too bloated/lacking features. As KB does actually function and is updated for security releases and bugs etc, I did put a hell of a lot of times into the plugins that I created and also that I took over.  But as you know it can get expensive over time in many ways.

> I can recognize you as one of the most active contributors, and I'm using myself several of your plugins - and BoardNotes was NOT made in the same way.
- thanks. I stopped coding for the last few months because I lost my job and although all my plugins (15+) have the sponsor link button on it, I never got more than $5 to fund my work.  The motivation to develop/adjust the way KB would do things rapidly went down after 367 days of non-stop learning and coding.  I'm not a developer by profession.  Having said that, I should get back into it.

> I'm using myself several of your plugins
- it would be great to know which ones you use?

> And I don't wanna be giving you constant headaches reviewing and merging massive rework changes.
- thats the way i have learnt github, git and KB development to be honest... your explaining of the changes in your commits are the excellent.

> So maybe consider giving the whole plugin a new beginning? Or even make it a fresh new plugin?
- for this plugin, I wouldn't know where to start as the JS and db side of things I am very weak on. I'm not good in php either in a complicated way.  I dont think the fresh new plugin would solve my issue or yours, as I feel the plugin was working to a KB version and then I took over.  It's not released for the exact reason.  I started working through it, recoding where I can through #1 and #2 and I know (in my coding style) there are quite a few things to do to keep it consistent with my other plugins.  I will work on that.

I can add you to the project if you like and then you can commit the changes directly, or you can kep it as it is - that's up to you.

So far I have merged your changes but not tested on my KB (I am one behind the latest version) but you said you side of the plugin is working so I think this plugin should be almost ready for the latest version as it is... just need to improve the coding style (which is no big deal for the end user) and add my style/templates to keep it a little consistent with my other plugins based on my [KanboardSkeletonPlugin](https://github.com/aljawaid/KanboardSkeletonPlugin).

What do you suggest?

**Comment by imfx77** [comment]
First of all - sorry for the looong answer ;)
Take your time.

> * yes it has been relativly stalled for over a year now

Yup, I see that. I also see you trying to raise some discussions on the matter with much of resistance from established KB contributors and no effect whatsoever. I also posted some issues and bugs and general workflow problems that got absolutely no attention ... So, yeah, I guess if I need smth on top of the KB framework - I have to rely on others' plugins or make it myself. Otherwise, I totally agree, there is no better alternative for me than KB, as a self-hosted PM software. I am not against paying for a commercial software - but there just isn't any that suits my needs, that is simple yet functional enough, that allows me control and custom enhancements. I've tried and evaluated lots of PM and CMS systems, and KB is still the best fit for me ... and furthermore I can contribute :)

> * I stopped coding for the last few months because I lost my job

I can totally understand that. Oddly enough, my situation is totally the opposite. I quit my job 2 years ago for various reason, mostly health issues induced by stress. Took me some time to recover my breath ... But since I have some savings to live on, and I am a long time game programmer, I have lately decided to go indie. Thus arranging my home studio was kinda the first step of it, and there is where KB kicks in. I'm no longer in a corporate environment which forces me to adopt a specific workflow and tools and management systems. And for the first time in my career I actually have resources time and will to engage with github and open source projects.

>  * it would be great to know which ones you use?

I am using your AutomaticActionUX, KanboardSupport, PluginManager and TagManager - all of them are a much convenience improvement and better UX to the original KB functionality. I considered for a while the CostControl but just couldn't find actual use for it in my current situation. I also used the ColorManager for a while trying to customize my visuals, but frankly I got overwhelmed by it and actually didn't achieve what I wanted. So just played with the CSS of the KB theme and customized it locally to get a semi transparent dark look.

> * thats the way i have learnt github, git and KB development to be honest...

Same thing here, I am a total newbie to all of it. Yet I have some solid background (I think) and the learning curve should be less steep for me.

>  your explaining of the changes in your commits are the excellent.

Work habits, of the good ones ;) The truth is I mostly do it for myself - having to review even my own commits a few months later isn't a trivial task - so a good comment is an invaluable aid, takes much less reverse engineering to understand what was done and why.

> * for this plugin, I wouldn't know where to start as the JS and db side of things I am very weak on. I'm not good in php either in a complicated way.

I guess I have a bit more general knowledge on those, but actually it takes me a lot of google-ing and reading tech docs and forums when I get to implement smth. Also it takes a great bunch of trial and errors. For the late part, I test things using at least 3x different browsers on 3x OS platforms, so to ensure it works/looks in the same way with a fair number of clients. Takes time, but nothing to do about that.

> just need to improve the coding style  (which is no big deal for the end user)

I saw some automatic linter warnings and errors when I merged your latest commit. Those indeed seem mostly mechanical to fix, except for the missing translations pack. It is good to have such scripts to make you keep up to some standards, I will have to learn how to use github actions myself.

> I dont think the fresh new plugin would solve my issue or yours

Unfortunately - no, it won't solve it. And I talk about that on the next comments.

> I need a post-it style system for 'temporary' notes related to a project/task
> I can add you to the project if you like and then you can commit the changes directly
> What do you suggest?

Well, imho, there are two ways to go:
1. keep the current plugin from releasing yet, see where it goes as features and enhancements, cleanup and polish the structure and code - then release it as a fairly mature product
2. start anew, lay down the proper skeleton, build on top of it, gradually and consistently

Pros and Cons of both:
1. The first approach will probably take more time to achieve the same results as the second. It will cost massive reworks to make it consistent clean and readable, in the meantime the reworks themselves will be messy and hard to follow. Hence, the project will need time to mature, and there isn't much sense in reviewing the intermediate changes, rather make a stable snapshot at some point, assume it as base - then make a one time effort to asses that base and fix everything, and then adopt a per feature approach if necessary to enhance it further.
2.  The second approach, starting anew - is definitely more appropriate if you think that the current plugin doesn't fit your purpose anyway. And here is, where it is very important to define what exactly is that purpose, what features should be considered, and how should they be presented to user UX-wise. This matters in the first case too, but there is a legacy there that drives you more or less into certain direction.

For example:
You state simply that you need a post-it style system for 'temporary' notes related to a project/task - which is a very vague definition - the current impl of the plugin can be assumed to satisfy that. But here go the details:
* a project related note or a task related note? those are different things - can you have both?
* what is considered a 'note'? just a simple text line or a rich text content?
* how do you manage and manipulate those notes? what actions can be performed?
* what properties or states do the notes have? what customizations? what presentation?
* etc etc etc ...

My interpretation for instance, went more towards TODO lists that are associated with projects or are custom projects on their own. For me it was important to be able easily reorder and sort them, to have visual aid by means of colorization by category, to be able to move them around projects, to easily update their contents and state, to have a one glance overview of ALL ... things like that ... I can't think of them all beforehand - it is the natural maturing process while using something and making it better on demand. Yet again, since it is an existing project, and not mine, I try keeping the design and the direction. E.g. I didn't find much sense in the 'report' feature, yet I updated it and made it consistent with my changes as it was already there. On the other hand, I reimplemented the NoteToTask feature from scratch as the 'jsonrpc client' approach seemed to me totally inappropriate given that the plugin works within the KB frame itself server-side. Just saying - legacy considerations.

So, to wrap up.
* Now that I've experimented enough on BoardNotes - I could keep my version which doesn't really fulfill my initial needs but I can make some use of it as a simpler TODO alternative to a kanban board. I will probably make a few more customization features and will stop evolving it. But if I make it for myself only it doesn't have to be polished, nor needs to be released. While making an official plugin assumes it will be useful and convenient for most users and hopefully will fill some gap they have with KB.
* So  the question here is if you and I envision the same/similar thing? And if that thing deserves to become a plugin? If NO - then I better stop messing up with your project and keep my own fork, if YES - then I would gladly contribute to your repo to develop it further.
* In the later case, if we consider implementing a fully finished product - we must define and sync purpose and requirements. Then we can take advantage of the both approaches. We can keep the current repo to experiment with and mature the features and UX, as it already has much of them working. And then we can consider either establishing it as a base and polishing it OR moving cleanly structured code into a new repo to get rid of the messy legacy.

I suggest we start at the beginning - by opening a discussion on this repo - to find out what are our wishes and expectations of such a plugin. To generally define purpose, features, UX. To see if we are on the same page about it all. As this will determine if we could work together or should better part ways.
Then depending on the goal and the current state of things next steps can be considered.

And my work on the project so far might be assumed just a warmup exercise :)

**Comment by imfx77** [comment]
btw, just to let you know
that I merged some more of my dev branch changes into [ imfx77/master](https://github.com/imfx77/kanboard-plugin-BoardNotes/tree/master)
mostly UX polish, CSS cleanup, handling text escaping properly, and of course more refactoring

as we have not yet considered which way the plugin shall go I won't bother you with a PR
(it's up to you if/when you decide to merge the above to your master)
nor will I proceed with the schema breaking changes that I have in mind for some further customization features

I will instead shift my attention for a while to some other KB issues that are currently boggling me
